### PR TITLE
add balanceTx

### PR DESCRIPTION
**Pull request by Geometer1729**
adds balanceTx, confirmBalanced and spend for https://github.com/mlabs-haskell/hedgehog-plutus-simple/issues/7

### CODE REVIEW

File: src/Hedgehog/Plutus/Tx.hs [hunk 1]
```diff
@@ -104,7 +112,125 @@ data ScriptPurpose
   In the future, hps should support custom balancers.
 -}
 balanceTx :: Model.Mock -> Tx 'Unbalanced -> Maybe (Tx 'Balanced)
-balanceTx = _
+balanceTx mock tx = do
+  let balanceingKeys = Map.keys $ Model.mockUsers mock
+  -- It probably doesn't make sense to balance a transaction
+  -- as every user at once in most cases
+  -- so we should probably make this an argument to balance
+  let valueOut = getValueOut tx
```
Comment (AriFordsham) [inline src/Hedgehog/Plutus/Tx.hs:120]: You should probably use `toLedgerTx` and  [`evaluateTransactionBalance`](https://github.com/input-output-hk/cardano-ledger/blob/e0947b0c2e99ef929d20593be03e199560392f42/eras/shelley/impl/src/Cardano/Ledger/Shelley/API/Wallet.hs#L496) to calculate balances

File: src/Hedgehog/Plutus/Tx.hs [hunk 2]
```diff
@@ -104,7 +112,125 @@ data ScriptPurpose
   In the future, hps should support custom balancers.
 -}
 balanceTx :: Model.Mock -> Tx 'Unbalanced -> Maybe (Tx 'Balanced)
-balanceTx = _
+balanceTx mock tx = do
+  let balanceingKeys = Map.keys $ Model.mockUsers mock
```
Comment (AriFordsham) [inline src/Hedgehog/Plutus/Tx.hs:116]: We don't need to know about keys, just take UTxOs until the balance is covered

**Review by AriFordsham — CHANGES_REQUESTED** [review]

File: src/Hedgehog/Plutus/Tx.hs (hunk 2 above)
Comment (Geometer1729) [reply src/Hedgehog/Plutus/Tx.hs:116]: I guess I still don't really understand why we would ever want to balance a transaction with utxos from all users so I wrote it to be easy to convert into balancing as a user or set of users.
If we're going to have happy path testing we will need to be able to balance as a single user or set of users, if we really need both balancing as every user just being a special case of balancing as a set of users feels like a good way to do that.

File: src/Hedgehog/Plutus/Tx.hs (hunk 1 above)
Comment (Geometer1729) [reply src/Hedgehog/Plutus/Tx.hs]: That took way longer than I expected lots of digging through https://github.com/input-output-hk/cardano-ledger ,https://github.com/input-output-hk/cardano-base and https://github.com/input-output-hk/plutus.

File: src/Hedgehog/Plutus/Tx.hs (hunk 2 above)
Comment (AriFordsham) [reply src/Hedgehog/Plutus/Tx.hs]: If the script relies on a specific PubKey, this must be described in the test data. If it does not, then we should test that the PubKey doesn't matter - even in happy path.

In my mind, 'happy path' means the script succeeds when it is supposed to - not just in limited cases when we expect it to.

File: src/Hedgehog/Plutus/Tx.hs [hunk 3]
```diff
@@ -106,11 +128,116 @@ data ScriptPurpose
...
+    protocol :: Model.PParams
+    protocol = Model.mockConfigProtocol mockConfig
+
+    mockConfig :: Model.MockConfig
+    mockConfig = Model.mockConfig $ mockchain context
+
+    body :: Class.IsCardanoTx era => TxBody era
+    body = Class.getTxBody $ toLedgerTx context tx
+
+maryToPlutus :: MV.MaryValue era -> Plutus.Value
```
Comment (AriFordsham) [inline src/Hedgehog/Plutus/Tx.hs:216]: It's in the back of my head that there's a library function for this.

**Review by AriFordsham — CHANGES_REQUESTED** [review]
Good work overall!

File: src/Hedgehog/Plutus/Tx.hs [hunk 4]
```diff
@@ -106,11 +128,116 @@ data ScriptPurpose
...
+                        , txOutReferenceScript = Nothing
+                        }
+                }
+    case checked of
+      Just tx -> pure tx
+      Nothing -> error "balanceTx produced an unbalanced tx"
+
+getBalance :: TxContext -> Tx b -> Maybe Value
+getBalance context tx = do
+  case protocol of
```
[the remark this answers is not in the document]
Comment (Geometer1729) [reply src/Hedgehog/Plutus/Tx.hs:202]: Thanks, It works with the inferred type but I can't seem to get it to work with a type signature.

File: src/Hedgehog/Plutus/Tx.hs (hunk 3 above)
Comment (Geometer1729) [reply src/Hedgehog/Plutus/Tx.hs]: I wouldn't be surprised but I couldn't find one for the life of me. Relatedly do you know how to get hoogle working for this repo?

**Comment by Geometer1729** [comment]
Sorry for the force push I accidentally pushed work from another PR here.

File: src/Hedgehog/Plutus/Tx.hs [hunk 5]
```diff
@@ -106,11 +131,140 @@ data ScriptPurpose
...
+            )
+    )
+    (== pkh)
+
+-- | as balanceTx but with additional predicates about which TxOuts and PubKeyHashs can be used
+balanceTxWhere ::
+  TxContext ->
+  Tx 'Unbalanced ->
+  (Plutus.TxOut -> Bool) ->
+  (Plutus.PubKeyHash -> Bool) ->
```
Comment (AriFordsham) [inline src/Hedgehog/Plutus/Tx.hs:157]: Do you have a motivation for a filter on the output, or is it just for completeness?

**Review by AriFordsham — APPROVED** [review]
LGTM!

File: src/Hedgehog/Plutus/Tx.hs (hunk 5 above)
Comment (Geometer1729) [reply src/Hedgehog/Plutus/Tx.hs:157]: Yeah, I needed a filter (or something similar) to make `balanceTxAsPubKey` return change to the expected user.

**Comment by Geometer1729** [comment]
I merged these prs in the wrong orders so I'll just manually merge the extra commits to ari/onion.

**Pull request merged**